### ISSUE TITLE
Strongbox is sometimes unable to see the latest version of an addon within 24 hours of release

### ISSUE THREAD

**Issue by AlexFolland** [ogri-la/strongbox#209]
I'm not sure why this is happening, but I got an email from CurseForge saying WorldFlightMap was updated.  I checked the WorldFlightMap page on CurseForge [here](https://www.curseforge.com/wow/addons/worldflightmap#c211) and found that it does indeed have a new release version as of 2 hours ago.  I tried to refresh my addon list in Strongbox and it didn't show the new version.  I manually overwrote my WorldFlightMap folder with the new version from the zip file, then refreshed Strongbox, and Strongbox says the old version is the new version.

![image](https://user-images.githubusercontent.com/601517/99621293-60e38700-29dc-11eb-8d86-dbe32f50bcc1.png)

**Comment by torkus** [comment]
hi @AlexFolland. Have you tried clearing the http cache? From the top menu:

> Cache -> Clear http cache

Requests are cached for 24hrs. This prevents the client (strongbox) hammering the server (curseforge, wowi, tukui, etc).

**Comment by torkus** [comment]
Did this solve your problem?

**Comment by AlexFolland** [comment]
It explains it.  Thanks!  It's good enough for most users.  Sorry for not responding for a few days.

As a user who wanted a specific addon checked for update, I was expecting right-clicking the addon and clicking "update" to bypass the cache.  Is that a possible new feature?  If not, that's fine.  Thanks again for your help and your awesome software.  I love it.

**Comment by torkus** [comment]
Improving the cache is on my [TODO](https://github.com/ogri-la/strongbox/blob/develop/TODO.md) list. I'll bump it so it gets some attention in the release *after* the one I'm prepping for at the moment (`3.1.0`). It's been a while since I looked at the HTTP side of things, a refresher would be good.

I'll close this ticket if you don't mind, I don't consider it a bug or an enhancement, but feel free to open other tickets if you have ideas.

**Comment by torkus** [comment]
I'm reopening this as an enhancement rather than a bug. 

When you say:

> I tried to refresh my addon list in Strongbox and it didn't show the new version. 

I imagine you were hitting `F5` or going `View` -> `Refresh` right? That's modeled directly on a web browser and when one does a 'refresh' in a browser they expect the page to be re-fetched regardless of the complex caching mechanisms in place. In the olden days of the web there was such a thing as [a 'hard' refresh](https://en.wikipedia.org/wiki/Wikipedia:Bypass_your_cache), where you would do a ctrl+F5 and it would bypass the cache. I can't remember the last time I've had to do that as browsers have gotten faster and cleverer but the principal is the same.

**Comment by AlexFolland** [comment]
Yes.  I have been using the "clear http cache" menu entry since then, if I need to force another check, which has worked for me.  I have been checking for new addons in the morning and again in the evening.  I'm sure checking with ~6-8 hours between checks is OK.

The thing I was really expecting to bypass the cache was checking for updates for a single addon, alone.  That couldn't hammer the server at all.  Maybe a maximum number of addons to check at once that bypasses the cache (like up to 5 or something like that) would make sense.

Cool that you'd reopen this as an enhancement.  I respect the work you do.

**Comment by AlexFolland** [comment]
I encountered something weird just now.  Even with HTTP cache cleared, strongbox isn't finding the latest Grid2 version.  I have r996 installed, but I see on both CurseForge and WowAce that Grid2 has r998 released 16 hours ago.  I specifically used "clear http cache" to get strongbox to update this, but it's not working.  Even "Re-install" just reinstalls r996.  I don't get why.

**Comment by torkus** [comment]
> Even with HTTP cache cleared, strongbox isn't finding the latest Grid2 version.

strongbox isn't talking to the website, it's talking to the CDN. Even if you clear the strongbox cache and make the request again, you're hitting the CDN immediately which is another type of cache. What is displayed on the website isn't always what is reflected on the CDN.

strongbox doesn't consider unstable or no-lib or 'other' types of releases. I'm not sure if it's possible, but the addon might have been released as unstable and then marked stable later on.

16 hours to invalidate a cache seems a bit much so something else might have been at play. I'm just guessing though, it's out of my hands once the request hits the internet.

**Comment by layday** [comment]
> I'm not sure if it's possible, but the addon might have been released as unstable and then marked stable later on.

Yes, that is possible, and the version can change too. `r988` of Grid2 was released as a beta (`r988-beta`) then re-tagged as `r988` some hours later.

**Comment by AlexFolland** [comment]
r998 was tagged "release".  I saw in my browser on both CurseForge and WowAce at that time.  I specifically checked for that.  I don't have definitive proof, but I'm 100% sure and you'll have to take my word for it.

**Comment by layday** [comment]
Thankfully, I log installed versions so we don't have to speculate:

```json
{
    "version": "r998",
    "install_time": "2020-11-30T22:21:47+00:00"
}
{
    "version": "r998-beta",
    "install_time": "2020-11-30T12:00:24+00:00"
}
```

**Comment by AlexFolland** [comment]
Those timestamps don't look right, unless you installed very late.  Here's what we can see on the CurseForge site for when the release package was uploaded:

![image](https://user-images.githubusercontent.com/601517/101154175-3c8dca00-35da-11eb-8542-49802b29822c.png)

**Comment by layday** [comment]
r988-beta was uploaded at 2020-11-30T03:11:46.91 UTC by CurseForge's [own reckoning](https://addons-ecs.forgesvc.net/api/v2/addon/15226/file/3126928).  I installed the beta some 9 hours later at 12:00 UTC and it was re-tagged as r988 and re-designated as stable sometime between noon and 22:21 UTC.  I don't know how this might all impact strongbox, but it's something I've observed happen often with Grid2.

**Comment by AlexFolland** [comment]
OK, great.  Nice link.  This is proof, then, that Strongbox had a bug here.  I commented here 16 hours after that upload timestamp, and I commented within fewer than 10 minutes of seeing the issue, meaning somehow strongbox was unable to get the newest version, even though my browser could see it from the same computer.

**Comment by torkus** [comment]
fixed in 3.2.0

refresh/F5 now clears the cache so there shouldn't be any cache hits from strongbox. Can't speak for the addon host or their setup though.
